util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+    :return: Token to access the database
+    """
+
+    data = {"key": api_key}
+    data = json.dumps(data).encode("utf-8")
+
+    req = request.Request(
+        auth_url, data=data, headers={"content-type": "application/json"}
+    )
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get token: {e}")
+
+    token = json.loads(response.read().decode("utf-8"))["access_token"]
+    return token
+
+
+def get_all_resources(
+    url: str, data_source: str, collection: str, database: str, token: str
+) -> List:
+    """
+    This function gets all the JSON objects for resources from the database
+    :param url: Database url to use Atlas data API
+    :param data_source: Data source name for the mongoDB database
+    :param collection: Collection name for the mongoDB database
+    :param database: Database name for the mongoDB database
+    :param token: Token to access the database
+
+    :return: List of JSON objects for resources
+    """
+
+    url = f"{url}/action/find"
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get resources: {e}")
+
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources: List, output: pathlib.Path):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def progress_hook(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize not in (None, -1):
+            t.total = tsize
+        displayed = t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+        return displayed
+
+    return update_to
+
+
+def download_resources(output: pathlib.Path):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    with open(output.joinpath("resources.json")) as f:
+        resources = json.load(f)
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filename = pathlib.Path(
+                *pathlib.Path(parse.urlsplit(url).path).parts[3:]
+            )
```

Comment:
This is really confusing, and if we end up keeping this I'd split it up into a line per function/step with comments explaining what each is doing. we're taking the URL, splitting it (?), getting the path, then casting it as a path, then getting the third part? ( din't know what part is and why "3" is significant). Is this just assuming our document is always 3-levels down a directory heirarchy? If so, please remember that the resource URL can be ANYTHING. It could be 65 directories down. There's no standard here.